modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/SqlRowHandler.java

```diff
@@ -0,0 +1,260 @@
+import org.apache.ignite.internal.sql.engine.exec.row.TypeSpec;
+import org.apache.ignite.internal.sql.engine.util.Commons;
+import org.apache.ignite.internal.sql.engine.util.TypeUtils;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Handler that uses a {@link RowWrapper wrapper} to operate with two types of row implementations.
+ *
+ * <ul>
+ *     <li>{@link ObjectsArrayRowWrapper Objects array}</li>
+ *     <li>{@link BinaryTupleRowWrapper Binary tuple}</li>
+ * </ul>
+ *
+ * <p>Each kind of rows is serialized to the same binary tuple format
+ * using the {@link #toByteBuffer(RowWrapper) toByteBuffer} method.
+ *
+ * <p>Factory methods {@link RowFactory#wrap(InternalTuple) wrap(InternalTuple)} and
+ * {@link RowFactory#create(ByteBuffer) create(ByteBuffer)} allow create rows without
+ * any additional conversions. But the fields in binary tuple must match the
+ * factory {@link RowSchema row schema}.
+ */
+public class SqlRowHandler implements RowHandler<RowWrapper> {
+    public static final RowHandler<RowWrapper> INSTANCE = new SqlRowHandler();
+
+    private SqlRowHandler() {
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public @Nullable Object get(int field, RowWrapper row) {
+        return row.get(field);
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public void set(int field, RowWrapper row, @Nullable Object val) {
+        row.set(field, val);
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public RowWrapper concat(RowWrapper left, RowWrapper right) {
+        int leftLen = left.columnsCount();
+        int rightLen = right.columnsCount();
+        List<TypeSpec> leftTypes = left.rowSchema().fields();
+        List<TypeSpec> rightTypes = right.rowSchema().fields();
+
+        Object[] values = new Object[leftLen + rightLen];
+        Builder schemaBuilder = RowSchema.builder();
+
+        for (int i = 0; i < leftLen; i++) {
+            values[i] = left.get(i);
+            schemaBuilder.addField(leftTypes.get(i));
+        }
+
+        for (int i = 0; i < rightLen; i++) {
+            values[leftLen + i] = right.get(i);
+            schemaBuilder.addField(rightTypes.get(i));
+        }
+
+        return new ObjectsArrayRowWrapper(schemaBuilder.build(), values);
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public RowWrapper map(RowWrapper row, int[] mapping, int offset) {
+        Object[] fields = new Object[mapping.length];
+
+        for (int i = 0; i < mapping.length; i++) {
+            fields[i] = row.get(mapping[i] + offset);
+        }
+
+        return new ObjectsArrayRowWrapper(row.rowSchema(), fields);
+    }
+
+    @Override
+    public int columnCount(RowWrapper row) {
+        return row.columnsCount();
+    }
+
+    @Override
+    public String toString(RowWrapper row) {
+        return String.valueOf(row);
+    }
+
+    @Override
+    public ByteBuffer toByteBuffer(RowWrapper row) {
+        return row.toByteBuffer();
```

Comment:
It would be better to provide implementation of `toString` method that is similar to ArrayRowHandler's. This one simply return `typeName@132` or something.

Reply:
Forgot about it - thanks for pointing on this.